### ISSUE TITLE
The Prison Theme

### ISSUE THREAD

**Issue by VezirChess** [official-stockfish/Stockfish#2117]
I will try to explain an Idea that Stockfish doesn't has.
This is THE PRİSON THEME.
The idea contains incarcerating opponent's pieces for draws or wins.
You can find some example positions in https://lichess.org/study/loM3CfBK
AlphaZero used this strategy a lot and this means Leela can use too.
I believe that If Stockfish developers adapts this idea in Stockfish, the elo gain will be more than 200...

**Comment by ddugovic** [comment]
Stockfish already has this: https://hxim.github.io/Stockfish-Evaluation-Guide/

https://github.com/official-stockfish/Stockfish/blob/6373fd56e90bc6114230a70cacad804248d955e2/src/evaluate.cpp#L134-L154
https://github.com/official-stockfish/Stockfish/blob/6373fd56e90bc6114230a70cacad804248d955e2/src/evaluate.cpp#L90-L106

**Comment by VezirChess** [comment]
Then why Stockfish can't solve that tactics in my study?
Why SF couldn't saw AlphaZero's moves or beat it.
There is something missing...

**Comment by VezirChess** [comment]
I am not talking about mobility. I am talking about Locking one of opponent's piece.
I am sure that Stockfish hasn't this strategy...

**Comment by ddugovic** [comment]
Duplicate of 7 other "fortress" issues:
https://github.com/official-stockfish/Stockfish/issues?utf8=✓&q=is%3Aissue+fortress

**Comment by VezirChess** [comment]
No it is not.
I have examined those issues and this is not same.
I am a international chess player and I can tell this theme is important. Marco, protonspring and others can you look at this issue?
Thank you

**Comment by ianfab** [comment]
One of the main problems with this thread is that you write 

> I believe that If Stockfish developers adapts this idea in Stockfish, the elo gain will be more than 200...

since after reading this sentence most people will perhaps just ignore the thread, since it indicates that you do not have much experience with engine development in practice. There is a huge difference between having a good idea for a pattern/technique that could be implemented into an engine in theory and finding an implementation that actually gains Elo in practice.

Most of the time the overhead for wasting time on evaluating a positional factor that does not apply in 99.9% of positions exceeds the gain from improving the evaluation of such 0.1% (or less) of positions. All the failed attempts on fortresses are typical examples.

Furthermore, such topics better fit into the [forum](https://groups.google.com/forum/#!forum/fishcooking) than here.

**Comment by acdemiralp** [comment]
Dude you are 1.3k. Anybody who has a brain is 1.3k. Take a look at code before you come up with "I'm the next hot shit" threads.

**Comment by VezirChess** [comment]
This is not about %0.01 positions.
For example, in many positions, the best answer for Nb6 is b3.
Because it blocks the knight. I think blocking is a very good technique and you should use it.
Best regards.

**Comment by VezirChess** [comment]
@acdemiralp What are you talking about? I just give them a technique

**Comment by ddugovic** [comment]
Stockfish already has bonuses for unblocked pieces.  They're even called ,CorneredBishop, LongDiagonalBishop MinorBehindPawn, Outpost, and RestrictedPiece:
https://github.com/official-stockfish/Stockfish/blob/6373fd56e90bc6114230a70cacad804248d955e2/src/evaluate.cpp#L134-L154

**Comment by VezirChess** [comment]
Whatever. You are developing Stockfish, not me.
Thanks anyway.

**Comment by VezirChess** [comment]
Just one more question.
How can I test my Patches?

**Comment by xoto10** [comment]
There's a wiki page describing how to do a first test:
https://github.com/ddugovic/Stockfish/wiki/Creating-my-first-test
It's a good idea to watch fishtest for a while to see how others do things

**Comment by Spill-The-Tea** [comment]
That lichess study was garbage... you are presenting well known fortress positions, including the Behting Study, and a modified version of Sir Roger Penrose's position as something novel without citing them. Not to mention, terrible commentary on alpha zero games.

For future reference, what you are calling "prison theme," is known by the chess world as fortress positions. The position shown from alpha zero games, are known as zugzwangs, which are different.

**Comment by VezirChess** [comment]
https://ailab.si/matej/doc/Detecting_Fortresses_in_Chess.pdf

**Comment by Spill-The-Tea** [comment]
Please read papers before citing them.
In the abstract alone, it states that this method is not viable for tournament play.
The "computational heuristic" they are prescribing in the paper is to avoid move pruning if the evaluation remains identical over large depths, and look at all possible moves to a certain depth x to reevaluate alternative moves. This just isn't feasible. However, you may be interested in matefinder, a stockfish fork found here, that does something similar, and can be useful for infinite analysis of difficult positions, such as fortresses: 

https://github.com/jhellis3/Stockfish/tree/mate_finder
